src/cpp/src/visual_language/inputs_embedder.cpp

```diff
@@ -112,19 +133,34 @@ class InputsEmbedder::IInputsEmbedder {
                 new_templated_chat_history = m_tokenizer.apply_chat_template(m_history, add_generation_prompt, chat_template_fallback);
             }
             ov::Tensor new_chat_tokens = m_tokenizer.encode(new_templated_chat_history).input_ids;
-            if (m_is_cache_empty) {
+            TokenizedInputs prev_chat_tokens = m_tokenizer.encode(m_templated_chat_history);
+
+            // some symbols combinations can be encoded by the tokenizer in different ways
+            // if we met sequence with such combination of symbols, we cannot correctly subtract the new history from the old history
+            // and find the difference as a prompt so let's check it out and use the whole history in this case
+            auto last_same_hist_token = 0;
+            if (!m_tokenized_chat_history.empty())
+                last_same_hist_token = ov::genai::utils::get_first_history_difference(prev_chat_tokens.input_ids, m_tokenized_chat_history);
+
+            if (m_tokenized_chat_history.empty()) {
                 encoded_input_ids = new_chat_tokens;
-                // after first `get_inputs_embeds` is called, we supposed LLM is inferred and cache is not empty
-                m_is_cache_empty = false;
+            } else if (!(last_same_hist_token == m_tokenized_chat_history.size() - 1)) {
+                m_to_remove_from_hist = m_tokenized_chat_history.size() - 1 - last_same_hist_token;
```

Comment:
maybe I misunderstand, but in case of the same histories `get_first_history_difference` will return history size, so should not we compare `last_same_hist_token == m_tokenized_chat_history.size()` (w/o `-1`) ?

Reply:
it's still not clear why do we need to subtract `1` here? Below we have similar code `new_chat_tokens.get_shape().at(1) - last_same_hist_token` w/o `-1` ..

Reply:
fixed in Fixed in https://github.com/openvinotoolkit/openvino.genai/pull/1254